interp/interp.go

```diff
@@ -0,0 +1,166 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package interp
+
+import (
+	"fmt"
+	"math"
+	"sort"
+
+	"gonum.org/v1/gonum/spatial/r1"
+)
+
+// Interpolator1D represents a 1-dimensional interpolator.
+// It interpolates a function over a defined closed interval range
+// (which can be infinite on both sides).
+type Interpolator1D interface {
+	// Interval returns the valid closed interval for interpolation.
+	Interval() r1.Interval
+
+	// ValueAt returns the interpolation value at x.
+	// It panics if x is not in the interpolation interval.
+	ValueAt(float64) float64
+}
+
+// Constant1D is a constant 1D interpolator,
+// It is defined over the range [-infinity, infinity].
+type Constant1D struct {
+	// Constant Y value.
+	Value float64
+}
+
+// Interval returns the valid closed interval for interpolation.
+func (ci Constant1D) Interval() r1.Interval {
+	return r1.Interval{Min: math.Inf(-1), Max: math.Inf(1)}
+}
+
+// ValueAt returns the interpolation value at x.
+func (ci Constant1D) ValueAt(x float64) float64 {
+	return ci.Value
+}
+
+// Linear1D is a piecewise linear 1-dimensional interpolator.
+type Linear1D struct {
+	// Interpolated X values.
+	xs []float64
+
+	// Interpolated Y data values, same len as ys.
+	ys []float64
+
+	// Slopes of Y between neighbouring X values. len(slopes) + 1 == len(xs) == len(ys).
+	slopes []float64
+}
+
+// NewLinear1D creates a new linear 1-dimensional interpolator.
```

Comment:
I don't think I like these New functions. I was expecting so see something like 

```
type Fitter interface {
   Fit(x, y []float64) error
}

type FittableInterpolator interface {
   Fitter
   Interpolator
}
```
In the simple case, this looks something like:
```
f := interp.PiecewiseLinear{}
err := f.Fit(xs, ys)
f.Predict(10)
```

But the power is that it enables easy building of routines that need a way of interpolating, but don't care about how.


You can imagine a function like 

```
func PredictAtPoints(locs []float64, f interp.FittableInterpolator, xdata, ydata []float64) ([]float64, error) {
    if err := f.Fit(xdata, ydata); err != nil{
         return nil, error
    }
    outs := make([]float64, len(locs))
    for i, v := range locs {
         outs[i] = f.Predict(v)
    }
    return outs, nil
}
```

And you would call it with something like

```
outs := make([]float64, 101)
floats.Span(outs, 0, 1)
scheme  := interp.Tikhonov{C: 10}
ys, err := mypkg.PredictAtPoints(outs, scheme, xdata, ydata)
```

You can then imagine building a loop to compare different predictors, etc.

Reply:
That's a good idea, thanks.

Reply:
One question: looking at `mat` package for example, I had the impression that when a numerical method fails, `gonum` prefers to panic than to return an error. But in your example `Fit` returns an error. What is the preferred way in `gonum`?